Core/src/Core/Filters/Algorithms/CalculateArrayHistogram.cpp

```diff
@@ -0,0 +1,315 @@
+#include "CalculateArrayHistogram.hpp"
+
+#include "Core/Filters/CalculateArrayHistogramFilter.hpp"
+
+#include "complex/Common/Range.hpp"
+#include "complex/DataStructure/DataArray.hpp"
+#include "complex/DataStructure/DataGroup.hpp"
+#include "complex/Utilities/ParallelDataAlgorithm.hpp"
+
+#include <chrono>
+
+using namespace complex;
+
+namespace
+{
+template <typename T>
+class GenerateHistogramFromData
+{
+public:
+  GenerateHistogramFromData(CalculateArrayHistogram& filter, const int32 numBins, const DataArray<T>& inputArray, Float64Array& histogram, std::atomic<usize>& overflow, float64 minRange = 0.0,
+                            float64 maxRange = 0.0)
+  : m_Filter(filter)
+  , m_NumBins(numBins)
+  , m_Min(minRange)
+  , m_Max(maxRange)
+  , m_InputArray(inputArray)
+  , m_Histogram(histogram)
+  , m_Overflow(overflow)
+  {
+  }
+  ~GenerateHistogramFromData() = default;
+
+  void generate(const usize start, const usize end) const
+  {
+    // tuple visulaization: Histogram = {(bin maximum, count), (bin maximum, count), ... }
+
+    float64 min = std::numeric_limits<float>::max();
+    float64 max = -1.0 * std::numeric_limits<float>::max();
+    if(m_Max == 0.0 && m_Min == 0.0)
+    {
+      const auto numOfElements = m_InputArray.getNumberOfComponents() * m_InputArray.getNumberOfTuples();
+      for(size_t i = 0; i < numOfElements; i++) // min and max in the input array
+      {
+        if(static_cast<float64>(m_InputArray[i]) > max)
+        {
+          max = static_cast<float64>(m_InputArray[i]);
+        }
+        if(static_cast<float64>(m_InputArray[i]) < min)
+        {
+          min = static_cast<float64>(m_InputArray[i]);
+        }
+      }
+      max = static_cast<float64>(max + 1); // ensure upper limit encapsulates max value
+      min = static_cast<float64>(min - 1); // ensure lower limit encapsulates min value
+    }
+    else
+    {
+      min = m_Min;
+      max = m_Max;
+    }
+
+    auto beginning = std::chrono::steady_clock::now();
+    const float64 increment = (max - min) / m_NumBins;
+    if(m_NumBins == 1) // if one bin, just set the first element to total number of points
+    {
+      m_Histogram[0] = max;
+      m_Histogram[1] = end;
+    }
+    else
+    {
+      for(usize i = start; i < end; i++)
+      {
+        m_Filter.updateThreadSafeProgress();
```

Comment:
Take a look at ApplyTransformationToGeometry filter and how the feedback from the parallel class is being performed. The problem I see here is that on _every_ iteration of this loop you are calling a function that has to get a lock which takes way too much time. This will artificially slow down the execution speed. ApplyTransformationToGeometry has a reasonable approach to this kind of feedback.